engine/server/sv_client.c

```diff
@@ -2119,6 +2119,780 @@ static qboolean SV_SendBuildInfo_f( sv_client_t *cl )
+
+	if( ent->v.solid )
+		SV_ClientPrintf( cl, "solid: %d\n", ent->v.solid );
+
+	SV_ClientPrintf( cl, "flags: 0x%x\n", ent->v.flags );
+	SV_ClientPrintf( cl, "spawnflags: 0x%x\n", ent->v.spawnflags );
+	return true;
+}
+
+/*
+===============
+SV_EntFire_f
+
+Perform some actions
+===============
+*/
+static qboolean SV_EntFire_f( sv_client_t *cl )
+{
+	edict_t	*ent = NULL;
+	int	i = 1, count = 0;
+	qboolean single; // true if user specified something that match single entity
+
+	if( Cmd_Argc() < 3 )
+	{
+		SV_ClientPrintf( cl, "Use ent_fire <index||pattern> <command> [<values>]\n"
+			"Use ent_fire 0 help to get command list\n" );
+		return false;
+	}
+
+	if( ( single = Q_isdigit( Cmd_Argv( 1 ) ) ) )
+	{
+		i = Q_atoi( Cmd_Argv( 1 ) );
+
+		if( i < 0 || i >= svgame.numEntities )
+			return false;
+
+		ent = EDICT_NUM( i );
+	}
+	else if( ( single = !Q_stricmp( Cmd_Argv( 1 ), "!cross" ) ) )
+	{
+		ent = SV_GetCrossEnt( cl->edict );
+
+		if (!SV_IsValidEdict(ent))
+			return false;
+
+		i = NUM_FOR_EDICT( ent );
+	}
+	else if( ( single = ( Cmd_Argv( 1 )[0] == '!') ) ) // Check for correct instanse with !(num)_(serial)
+	{
+		const char *cmd = Cmd_Argv( 1 ) + 1;
+		i = Q_atoi( cmd );
+
+		while( Q_isdigit( cmd )) cmd++;
+
+		if( *cmd++ != '_' )
+			return false;
+
+		if( i < 0 || i >= svgame.numEntities )
+			return false;
+
+		ent = EDICT_NUM( i );
+		if( ent->serialnumber != Q_atoi( cmd ) )
+			return false;
+	}
+	else
+	{
+		i = svgame.globals->maxClients + 1;
+	}
+
+	for( ; ( i <  svgame.numEntities ) && ( count < sv_enttools_maxfire.value ); i++ )
+	{
+		ent = EDICT_NUM( i );
+		if( !SV_IsValidEdict( ent ))
+		{
+			// SV_ClientPrintf( cl, PRINT_LOW, "Got invalid entity\n" );
+			if( single )
+				break;
+			continue;
+		}
+
+		// if user specified not a number, try find such entity
+		if( !single )
+		{
+			if( !Q_stricmpext( Cmd_Argv( 1 ), STRING( ent->v.targetname ) ) && !Q_stricmpext( Cmd_Argv( 1 ), STRING( ent->v.classname ) ))
+				continue;
+		}
+
+		SV_ClientPrintf( cl, "entity %i\n", i );
+
+		count++;
+
+		if( !Q_stricmp( Cmd_Argv( 2 ), "health" ) )
+			ent->v.health = Q_atoi( Cmd_Argv ( 3 ) );
+		else if( !Q_stricmp( Cmd_Argv( 2 ), "gravity" ) )
+			ent->v.gravity = Q_atof( Cmd_Argv ( 3 ) );
+		else if( !Q_stricmp( Cmd_Argv( 2 ), "movetype" ) )
+			ent->v.movetype = Q_atoi( Cmd_Argv ( 3 ) );
+		else if( !Q_stricmp( Cmd_Argv( 2 ), "solid" ) )
+			ent->v.solid = Q_atoi( Cmd_Argv ( 3 ) );
+		else if( !Q_stricmp( Cmd_Argv( 2 ), "rename" ) )
+			ent->v.targetname = ALLOC_STRING( Cmd_Argv ( 3 ) );
+		else if( !Q_stricmp( Cmd_Argv( 2 ), "settarget" ) )
+			ent->v.target = ALLOC_STRING( Cmd_Argv ( 3 ) );
+		else if( !Q_stricmp( Cmd_Argv( 2 ), "setmodel" ) )
+			SV_SetModel( ent, Cmd_Argv( 3 ) );
+		else if( !Q_stricmp( Cmd_Argv( 2 ), "set" ) )
+		{
+			char keyname[MAX_STRING];
+			char value[MAX_STRING];
+			KeyValueData	pkvd;
+			if( Cmd_Argc() != 5 )
+				return false;
+
+			pkvd.szClassName = (char*)STRING( ent->v.classname );
```

Comment:
There really needed such cast?

Reply:
Yes, because pkvd structure fields is `char *` type. Maybe we can replace it type to `const char *`, but I dunno.

Reply:
Ah... ok.
Anyway we can change type:
https://github.com/FWGS/xash3d-fwgs/blob/master/engine/eiface.h#L290=

Reply:
I think changing structure type is not in this PR scope :)